app/frontend/src/features/auth/signup/Signup.test.tsx

```diff
@@ -0,0 +1,235 @@
+    );
+    userEvent.click(hostingStatusItem);
+
+    userEvent.click(screen.getByLabelText(FEMALE));
+    userEvent.click(screen.getByLabelText(SIGN_UP_TOS_ACCEPT));
+
+    userEvent.click(screen.getByRole("button", { name: SIGN_UP }));
+
+    await waitFor(() => {
+      expect(screen.getByText(QUESTIONS_OPTIONAL)).toBeVisible();
+    });
+
+    userEvent.click(screen.getByRole("button", { name: SUBMIT }));
+
+    expect(await screen.findByTestId("dashboard")).toBeVisible();
+  });
+
+  it.each`
+    hasAuthRes | needBasic | needAccount | needFeedback | needVerifyEmail | expectedForm
+    ${false}   | ${true}   | ${true}     | ${true}      | ${true}         | ${"basic"}
+    ${false}   | ${false}  | ${true}     | ${true}      | ${true}         | ${"account"}
+    ${false}   | ${false}  | ${true}     | ${false}     | ${true}         | ${"account"}
+    ${false}   | ${false}  | ${true}     | ${false}     | ${false}        | ${"account"}
+    ${false}   | ${false}  | ${false}    | ${true}      | ${true}         | ${"feedback"}
+    ${false}   | ${false}  | ${false}    | ${true}      | ${false}        | ${"feedback"}
+    ${false}   | ${false}  | ${false}    | ${false}     | ${true}         | ${"email"}
+    ${false}   | ${false}  | ${false}    | ${false}     | ${false}        | ${"error"}
+    ${true}    | ${false}  | ${false}    | ${false}     | ${false}        | ${"dashboard"}
+  `(
+    "displays $expectedForm form given state: $hasAuthRes $needBasic $needAccount $needFeedback $needVerifyEmail",
+    async (data) => {
+      const state: SignupFlowRes.AsObject = {
+        ...data,
+        hasAuthRes: undefined,
+        expectedForm: undefined,
+        flowToken: "token",
+        authRes: data.hasAuthRes ? { userId: 1, jailed: false } : undefined,
+      };
+      window.localStorage.setItem("auth.flowState", JSON.stringify(state));
+
+      if (data.expectedForm === "error") {
+        mockConsoleError();
+        expect(() =>
+          render(<View />, {
+            wrapper: getHookWrapperWithClient({
+              initialRouterEntries: [signupRoute],
+            }).wrapper,
+          })
+        ).toThrow();
+        return;
+      }
+
+      render(<View />, {
+        wrapper: getHookWrapperWithClient({
+          initialRouterEntries: [signupRoute],
+        }).wrapper,
+      });
+
+      switch (data.expectedForm) {
+        case "basic":
+          expect(screen.getByLabelText(EMAIL_LABEL)).toBeVisible();
+          break;
+        case "account":
+          expect(screen.getByLabelText(USERNAME)).toBeVisible();
+          break;
+        case "feedback":
+          expect(screen.getByText(QUESTIONS_OPTIONAL)).toBeVisible();
+          break;
+        case "email":
+          expect(screen.getByText(SIGN_UP_AWAITING_EMAIL)).toBeVisible();
+          break;
+        case "redirect":
+          expect(await screen.findByText(SIGN_UP_REDIRECT)).toBeVisible();
+          break;
+      }
+    }
+  );
```

Comment:
Considering the label will say something like "display basic form given state: false false false true false", it is a bit opaque to read and follow. Also there's a fair bit of logic in the test itself which themselves can't be tested, so not sure this is best use case for `it.each` here.

I think it'd be a lot clearer for these to be split out into separate test cases, even if it means a bit of repetition, but at least the test label will read more clearly than a list of booleans...

Reply:
How strongly do you feel about it? I know it seems a bit weird with the label but I thought if there's a problem with the test, you will be looking at this part of the file anyway, and the label makes it very clear which case is being tested. It's going to be at least 100 extra lines to set up each of these cases individually

Reply:
I think the extra lines because of repetition are a good sacrifice for better readability in test code. If any of these fail, there's big productivity value in being able to figure out which one failed from the test output without having to come back here to work out which branch caused the problem imo.

Aside from the labels, the extra logic within the test adds a layer of abstraction which makes them less straightforward to follow, as this is essentially multiple tests squashed into one when each really deserve its own test. [This blog post](https://testing.googleblog.com/2014/07/testing-on-toilet-dont-put-logic-in.html) sums up my thinking more coherently than what I can say here I feel :p